src/stdlib_selection.fypp

```diff
@@ -0,0 +1,275 @@
+                    end if
+                  end do
+                  call swap(array(left),array(hi))
+                  iPivot=hi
+              end subroutine
+      end subroutine
+    #:endfor
+  #:endfor
+
+
+  #:for arraykind, arraytype in ARRAY_KINDS_TYPES
+    #:for intkind, inttype in INT_KINDS_TYPES
+      #:set name = rname("arg_select", 1, arraytype, arraykind, intkind)
+      subroutine ${name}$(a, indx, k, kth_smallest, left, right)
+          !! arg_select - find the index of the k-th smallest entry in `a(:)`
+          !!
+          !! Partly derived from the "Coretran" implementation of 
+          !! quickSelect by Leon Foks, https://github.com/leonfoks/coretran
+          !!
+          ${arraytype}$, intent(in) :: a(:)
+              !! Array in which we seek the k-th smallest entry.
+          ${inttype}$, intent(inout) :: indx(:)
+              !! Array of indices into `a(:)`. Must contain each integer
+              !! from `1:size(a)` exactly once. On output it will be partially
+              !! sorted such that
+              !! `all( a(indx(1:(k-1)))) <= a(indx(k)) ) .AND.
+              !!  all( a(indx(k))  <= a(indx( (k+1):size(a) )) )`.
+          ${inttype}$, intent(in) :: k
+              !! We want index of the k-th smallest entry. E.G. `k=1` leads to
+              !! `a(kth_smallest) = min(a)`, and `k=size(a)` leads to
+              !! `a(kth_smallest) = max(a)`
+          ${inttype}$, intent(out) :: kth_smallest
+              !! On output contains the index with the k-th smallest value of `a(:)`
+          ${inttype}$, intent(in), optional :: left, right
+              !! If we know that:
+              !!  the k-th smallest entry of `a` is in `a(indx(left:right))`
+              !! and also that:
+              !!  `maxval(a(indx(1:(left-1)))) <= minval(a(indx(left:right)))`
+              !! and:
+              !!  `maxval(a(indx(left:right))) <= minval(a(indx((right+1):size(a))))`
+              !! then one or both bounds can be specified to reduce the search
+              !! time. These constraints are available if we have previously
+              !! called the subroutine with a different `k` (due to the way that
+              !! `indx(:)` becomes partially sorted, see documentation for `indx(:)`).
+
+          ${inttype}$ :: l, r, mid, iPivot
+          integer, parameter :: ip = ${intkind}$
+
+          l = 1_ip
+          if(present(left)) l = left
+          r = size(a, kind=ip)
+          if(present(right)) r = right
+
+          if(size(a) /= size(indx)) then
+              error stop "arg_select must have size(a) == size(indx)"
+          end if
+
+          if(k < 1_ip .or. k > size(a, kind=ip) .or. l > r .or. l < 1_ip .or. &
+              r > size(a, kind=ip)) then
+              error stop "arg_select must have 1 <= k <= size(a), and 1 <= left <= right <= size(a)";
+          end if
+
+          do while(.true.)
+              mid = (l+r)/2_ip ! Deliberate integer division
```

Comment:
This line might cause an integer overflow if the arrays being sorted are extremely large. A fix is to rewrite the expression as `l + ((r - l)/2_ip)`. It's a known bug in sorting routines: 
* [Extra, Extra - Read All About It: Nearly All Binary Searches and Mergesorts are Broken](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html)

Reply:
Good catch. I suggest to add a test that would trigger this possible issue.

Reply:
There was a HackerNews thread about it: https://news.ycombinator.com/item?id=1130463

The suggested fix doesn't really eliminate the bug, but just extend the region of validity slightly. E.g. with the default (signed) integers (32-bits), you will get wrap-around if the sum `l + r` exceeds 2147483637 = 2^32 - 1. (An array of 4-byte reals this size would have 128 GiB if I'm not mistaken.)

This does make me wonder how does one deal with an array with custom bounds like:
```
real :: a(-2147483637:2147483637)
```
where the entire array is addressable using default integers.

Reply:
Great catch!

Reply:
A commit just now uses the more robust `mid` calculation, and also adds new tests (that failed with the old method, but work now).